File: adapters/dotnet/src/AuthNinja.AspNetCore/Auth/WebAuthnChallengeStore.cs
using Fido2NetLib;

namespace AuthNinja.AspNetCore.Auth;

internal enum WebAuthnChallengeKind
{
    Register,
    Login,
}

internal sealed class StoredWebAuthnChallenge
{
    public required WebAuthnChallengeKind Kind { get; init; }

    public Guid? UserId { get; init; }

    public required DateTimeOffset ExpiresAt { get; init; }

    public CredentialCreateOptions? CreateOptions { get; init; }

    public AssertionOptions? AssertionOptions { get; init; }
}

internal sealed class InMemoryWebAuthnChallengeStore
{
    private readonly Dictionary<string, StoredWebAuthnChallenge> _entries = new();

    public void Set(
        string challenge,
        WebAuthnChallengeKind kind,
        Guid? userId,
        TimeSpan ttl,
        DateTimeOffset now,
        CredentialCreateOptions? createOptions = null,
        AssertionOptions? assertionOptions = null)
    {
        _entries[challenge] = new StoredWebAuthnChallenge
        {
            Kind = kind,
            UserId = userId,
            ExpiresAt = now.Add(ttl),
            CreateOptions = createOptions,
            AssertionOptions = assertionOptions,
        };
    }

    public StoredWebAuthnChallenge? Consume(string challenge, WebAuthnChallengeKind kind, DateTimeOffset now)
    {
        if (!_entries.TryGetValue(challenge, out var entry))
        {
            return null;
        }

        _entries.Remove(challenge);

        if (entry.Kind != kind || now > entry.ExpiresAt)
        {
            return null;
        }

        return entry;
    }
}

File: packages/next/src/auth/webauthn-challenge-store.ts
/** WebAuthn ceremony type stored with the challenge. */
export type WebAuthnChallengeKind = "register" | "login";

export type WebAuthnChallengeRecord = {
  kind: WebAuthnChallengeKind;
  /** Set for registration and email-scoped login; resolved from credential on discoverable login. */
  userId?: string;
  expiresAt: number;
};

export type WebAuthnChallengeStore = {
  set(
    challenge: string,
    record: Omit<WebAuthnChallengeRecord, "expiresAt">,
    ttlMs: number,
    now?: Date,
  ): void | Promise<void>;
  consume(
    challenge: string,
    kind: WebAuthnChallengeKind,
    now?: Date,
  ): WebAuthnChallengeRecord | undefined | Promise<WebAuthnChallengeRecord | undefined>;
};

type StoredChallenge = WebAuthnChallengeRecord;

/** In-memory WebAuthn challenge store with short TTL (single-process dev default). */
export class InMemoryWebAuthnChallengeStore implements WebAuthnChallengeStore {
  private readonly entries = new Map<string, StoredChallenge>();

  set(
    challenge: string,
    record: Omit<WebAuthnChallengeRecord, "expiresAt">,
    ttlMs: number,
    now: Date = new Date(),
  ): void {
    this.entries.set(challenge, {
      ...record,
      expiresAt: now.getTime() + ttlMs,
    });
  }

  consume(
    challenge: string,
    kind: WebAuthnChallengeKind,
    now: Date = new Date(),
  ): WebAuthnChallengeRecord | undefined {
    const entry = this.entries.get(challenge);

    if (!entry || entry.kind !== kind || now.getTime() > entry.expiresAt) {
      this.entries.delete(challenge);
      return undefined;
    }

    this.entries.delete(challenge);
    return entry;
  }
}
